### PR TITLE
KAFKA-3272: Add debugging options to kafka-run-class.sh so we can easily run remote debugging

### PR DESCRIPTION
**Pull request by christian-posta**
…able remote debugging to Kafka tools scripts

### CODE REVIEW

**Comment by gwenshap** [comment]
Yes! Love this. Maybe we need a flag for suspend=y? I find that I need it more often than not.

**Comment by christian-posta** [comment]
yah good point @gwenshap! tbh, i always use that flag. let me add it as an option.

File: bin/kafka-run-class.sh [hunk 1]
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
   KAFKA_OPTS=""
 fi
 
+# Set Debug options if enabled
+DEFAULT_JAVA_DEBUG_PORT="5005"
+if [ "x$JAVA_DEBUG_PORT" = "x" ]; then
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: Can we change this to match the other "presence" checks in the script:

```
if [ -z "$JAVA_DEBUG_PORT" ]
```

Comment (granthenke) [inline bin/kafka-run-class.sh]: Should this move inside the section below  the `if KAFKA_DEBUG` check?

File: bin/kafka-run-class.sh [hunk 2]
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
...
+    JAVA_DEBUG_PORT="$DEFAULT_JAVA_DEBUG_PORT"
+fi
+
+DEFAULT_JAVA_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=${DEBUG_SUSPEND_FLAG:-n},address=$JAVA_DEBUG_PORT"
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: Should this move inside the section below the `if KAFKA_DEBUG` check?

File: bin/kafka-run-class.sh [hunk 3]
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
...
+if [ "x$KAFKA_DEBUG" != "x" ]; then
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: Same as above, use `-z`

File: bin/kafka-run-class.sh [hunk 4]
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
...
+    # Use the defaults if JAVA_DEBUG_OPTS was not set
+    if [ "x$JAVA_DEBUG_OPTS" = "x" ]; then
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: Same as above, use `-z`

File: bin/kafka-run-class.sh (hunk 1 above)
Comment (christian-posta) [inline bin/kafka-run-class.sh]: yah let me refactor that, thanks for the review!

File: bin/kafka-run-class.sh (hunk 2 above)
Comment (christian-posta) [inline bin/kafka-run-class.sh]: +1 will change

**Comment by granthenke** [comment]
Thanks for the patch, just a few small comments. 

Could you change the title of the PR to: "KAFKA-3272: Add debugging options to kafka-run-class.sh so we can easily run remote debugging" so that it links to the jira?

Would you be interested in making the Windows .bat script match too?

**Comment by christian-posta** [comment]
@granthenke force-updated the commit with the changes you suggested. thanks!

let me poke around at the windows script.

File: bin/kafka-run-class.sh [hunk 5]
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
...
+        JAVA_DEBUG_OPTS="$DEFAULT_JAVA_DEBUG_OPTS"
+    fi
+
+    KAFKA_OPTS="$JAVA_DEBUG_OPTS $KAFKA_OPTS"
+    warn "Enabling Java debug options: $JAVA_DEBUG_OPTS"
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: Not sure you need the `warn` line

**Comment by gwenshap** [comment]
@granthenke @christian-posta Do you think we should merge now and follow up on windows later? Or add windows to this PR and avoid diverging the scripts?

File: bin/kafka-run-class.sh [hunk 6]
```diff
@@ -115,6 +115,28 @@ if [ -z "$KAFKA_OPTS" ]; then
...
+    # Use the defaults if JAVA_DEBUG_OPTS was not set
+    DEFAULT_JAVA_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=${DEBUG_SUSPEND_FLAG:-n},address=$JAVA_DEBUG_PORT"
+    if [ -z "$JAVA_DEBUG_OPTS" ]; then
+        JAVA_DEBUG_OPTS="$DEFAULT_JAVA_DEBUG_OPTS"
+    fi
+
+
+
+    KAFKA_OPTS="$JAVA_DEBUG_OPTS $KAFKA_OPTS"
+    warn "Enabling Java debug options: $JAVA_DEBUG_OPTS"
```
Comment (granthenke) [inline bin/kafka-run-class.sh]: I am not familiar with `warn`. Do we need this?

**Comment by granthenke** [comment]
@gwenshap I am okay, if we merge and open a jira for windows. I just had one comment on the usage of `warn`. Otherwise LGTM.

**Comment by christian-posta** [comment]
@granthenke @gwenshap i've changed the warn to just "echo" and added the windows equivalent. 

windows batch file expansion rules.. learn something new everyday!

**Comment by gwenshap** [comment]
...if there was only a place to document those cool new features :)

Actually maybe we don't have a good place for that yet. I'll commit and open a JIRA for command line docs.

**Comment by gwenshap** [comment]
LGTM.

**Comment by christian-posta** [comment]
@gwenshap shall I open another JIRA for the patch for the windows batch files? Already have it in this branch...

**Pull request closed without merging**